vernac/comInductive.ml

```diff
@@ -138,17 +119,37 @@ let interp_ind_arity env sigma ind =
   in
   sigma, (t, pseudo_poly, impls)
 
-let interp_cstrs env sigma impls mldata arity ind =
+(* ind_rel is the Rel for this inductive in the context without params.
+   n is how many arguments there are in the constructor. *)
+let model_conclusion env sigma ind_rel params n arity_indices =
+  let model_head = EConstr.mkRel (n + Context.Rel.length params + ind_rel) in
+  let model_params = Context.Rel.to_extended_vect EConstr.mkRel n params in
+  let sigma,model_indices =
+    List.fold_right
+      (fun (_,t) (sigma, subst) ->
+        let t = EConstr.Vars.substl subst (EConstr.Vars.liftn n (List.length subst + 1) t) in
+        let sigma, c = Evarutil.new_evar env sigma t in
+        sigma, c::subst)
+      arity_indices (sigma, []) in
+  sigma, EConstr.mkApp (EConstr.mkApp (model_head, model_params), Array.of_list (List.rev model_indices))
+
+let interp_cstrs env (sigma, ind_rel) impls params ind arity =
   let cnames,ctyps = List.split ind.ind_lc in
-  (* Complete conclusions of constructor types if given in ML-style syntax *)
-  let ctyps' = List.map2 (complete_conclusion mldata) cnames ctyps in
+  let arity_indices, cstr_sort = Reductionops.splay_arity env sigma arity in
   (* Interpret the constructor types *)
+  let interp_cstr sigma ctyp =
+    let sigma, (ctyp', cimpl) = interp_type_evars_impls env sigma ~impls ctyp in
+    let ctx, concl = Reductionops.splay_prod_assum env sigma ctyp' in
+    let concl_env = EConstr.push_rel_context ctx env in
+    let sigma, model = model_conclusion concl_env sigma ind_rel params (Context.Rel.length ctx) arity_indices in
+    (* unify the expected with the provided conclusion *)
+    let sigma = Option.default sigma (Evarconv.conv concl_env sigma concl model) in
```

Comment:
Now that's really strange, you're ignoring a potential unification failure of the conclusion and your model here. By construction they should unify I suppose, so you should rather raise an anomaly in case conversion fails here.

Reply:
We are still before the kernel here; the user may have given an ill-formed constructor. Originally, I caught the error and gave a `user_error`, but the kernel already has error-catching code for this, so it seemed simpler to ignore the issue here and let the kernel catch it later (The test-suite also checks this error message: https://github.com/coq/coq/blob/master/test-suite/output/Inductive.out).
I added a comment to the code making this explicit.